src/frontend/src/optimizer/plan_node/logical_agg.rs

```diff
@@ -184,60 +196,94 @@ impl LogicalAgg {
+    fn all_local_aggs_are_stateless(&self, stream_input_append_only: bool) -> bool {
+        self.agg_calls().iter().all(|c| {
+            matches!(c.agg_kind, AggKind::Sum | AggKind::Count)
+                || (matches!(c.agg_kind, AggKind::Min | AggKind::Max) && stream_input_append_only)
+        })
+    }
 
-        // some agg function can not rewrite to 2-phase agg
-        // we can only generate stand alone plan for the simple agg
-        let all_agg_calls_can_use_two_phase = self.can_two_phase_agg();
-        if !all_agg_calls_can_use_two_phase {
-            return gen_single_plan(stream_input);
+    /// Generally used by two phase hash agg.
+    /// If input dist already satisfies hash agg distribution,
+    /// it will be more expensive to do two phase agg, should just do shuffle agg.
+    pub(crate) fn hash_agg_dist_satisfied_by_input_dist(&self, input_dist: &Distribution) -> bool {
+        let required_dist =
+            RequiredDist::shard_by_key(self.input().schema().len(), self.group_key());
+        input_dist.satisfies(&required_dist)
+    }
+
+    /// Generates distributed stream plan.
+    fn gen_dist_stream_agg_plan(&self, stream_input: PlanRef) -> Result<PlanRef> {
+        let input_dist = stream_input.distribution();
+        debug_assert!(*input_dist != Distribution::Broadcast);
+
+        // Shuffle agg
+        // If we have group key, and we won't try two phase agg optimization at all,
+        // we will always choose shuffle agg over single agg.
+        if !self.group_key().is_empty() && !self.must_try_two_phase_agg() {
+            return self.gen_shuffle_plan(stream_input);
         }
 
-        // stateless 2-phase simple agg
-        // can be applied on stateless simple agg calls with input distributed by any shard
-        let all_local_are_stateless = self.agg_calls().iter().all(|c| {
-            matches!(c.agg_kind, AggKind::Sum | AggKind::Count)
-                || (matches!(c.agg_kind, AggKind::Min | AggKind::Max) && input_append_only)
+        // Standalone agg
+        // If no group key, and cannot two phase agg, we have to use single plan.
+        if self.group_key().is_empty() && !self.can_two_phase_agg() {
+            return self.gen_single_plan(stream_input);
+        }
+
+        debug_assert!(if !self.group_key().is_empty() {
+            self.must_try_two_phase_agg()
+        } else {
+            self.can_two_phase_agg()
         });
-        if all_local_are_stateless
+
+        // Stateless 2-phase simple agg
+        // can be applied on stateless simple agg calls,
+        // with input distributed by [`Distribution::AnyShard`]
+        if self.group_key().is_empty()
+            && self.all_local_aggs_are_stateless(stream_input.append_only())
             && input_dist.satisfies(&RequiredDist::AnyShard)
-            && self.group_key().is_empty()
         {
             return self.gen_stateless_two_phase_streaming_agg_plan(stream_input);
         }
 
```

Comment:
```suggestion
        let (stream_input, input_dist) = if self.group_key().is_empty() && self.must_try_two_phase_agg() {
                let stream_input = RequiredDist::ShardByKey(stream_input.logical_pk().into()).enforce_if_not_satisfies(stream_input);
                (stream_input, stream_input.distribution())
        } else {
                (stream_input, input_dist) 
        };
```

Reply:
Why do we enforce `logical_pk` distribution on stream_input? I thought as long as it follows `HashShard` / `UpstreamHashShard` it is fine, regardless of whether `self.group_key()` is empty or not? Since we will do shuffle before the global agg step.

Any example query you can share? I'd like to use it for testcase as well.

Reply:
Sorry I got it wrong and here I just want to handle `SomeShard` distribution here. 
The  `SomeShard` distribution describes that the data is distributed in different partitions but does not obey any consistent hash strategy in our system. For example: We know a Kafka source fragment is distributed the kafka's split but it does not follow the consistent distribution in our system. 
Though we need to try our best not to use `SomeShard` in our system. And after I consider it case by case, I think we do not need to handle it here( all source fragment is append-only and it could be handled by stateless 2-phase agg). I think we still should need handle it here. Because when user use `RW_FORCE_TWO_PHASE_AGG=true` they mean they do not want to get a stand-alone plan. So we must to make sure that the plan is processed and distributed with 2-phase agg. So here we convert the `SomeShard` to `HashShard` by a forcing exchange. Using logical pk is just make sure that all operations on the same key can be shuffled into one node... it should not be required but here we need a distribution key anyway and maybe the primary key can take less skewness(not very sure).

Reply:
Makes sense, thanks for the explanation!